### ISSUE TITLE
Format Json Output

### ISSUE THREAD

**Issue by lilysikes** [samuelcolvin/pydantic#1315]
# Question

Output of `python -c "import pydantic.utils; print(pydantic.utils.version_info())"`:
```
pydantic version: 1.4
            pydantic compiled: True
                 install path: /opt/conda/lib/python3.6/site-packages/pydantic
               python version: 3.6.8 |Anaconda, Inc.| (default, Dec 30 2018, 01:22:34)  [GCC 7.3.0]
                     platform: Linux-4.19.76-linuxkit-x86_64-with-debian-stretch-sid
     optional deps. installed: ['typing-extensions', 'email-validator']
```
For HttpUrl and IPvAnyAddress I would like to format the json output. 
For the url I get the following:
```
from pydantic import BaseModel, HttpUrl

class Test(BaseModel):
    url: HttpUrl

m = Test(url="http://www.example.com/123/abc?123#abc")
m.json()
# '{"url": "http://www.example.com/123/abc?123#abc"}'
```
Instead I would like the json to include all of the properties of the url. Something like:
```
'{"url":{"scheme": "http", "host": "www.example.com",  "tld": "com", "host_type": "domain", "path": "/123/abc", "query": "123", "fragment": "abc"}}'
```
For the IPvAnyAddress I would like to return some of the properties that the ipaddress provides. For example:
```
from pydantic import BaseModel, IPvAnyAddress

class Test(BaseModel):
    ip: IPvAnyAddress

m = Test(ip="255.255.255.0")
m.json()
# '{"ip": "255.255.255.0"}'
```
Instead I would like something like:
```
'{"ip": {"compressed": "255.255.255.0", "version": 4, "is_global": False, "is_private": True, "is_multicast": False}}'
```
Is either of these things possible and if so how?

**Comment by StephenBrown2** [comment]
Huh, I would have expected this to be possible by stealing some logic from the repr function on AnyUrl for a custom `json_encoder`, but it doesn't work as I would expect either:
```
In [1]: from pydantic import BaseModel, HttpUrl 
   ...:  
   ...: def dump_attrs(v): 
   ...:     return {n: getattr(v, n) for n in v.__class__.__slots__ if getattr(v, n) is not None} 
   ...:  
   ...: class MyModel(BaseModel): 
   ...:     url: HttpUrl 
   ...:  
   ...:     class Config: 
   ...:         json_encoders = { 
   ...:             HttpUrl: dump_attrs 
   ...:         } 
   ...:  
   ...: m = MyModel(url="http://www.example.com/123/abc?123#abc")                                                                                         

In [2]: m.json()                                                                                                                                          
Out[2]: '{"url": "http://www.example.com/123/abc?123#abc"}'

In [3]: dump_attrs(m.url)                                                                                                                                 
Out[3]: 
{'scheme': 'http',
 'host': 'www.example.com',
 'tld': 'com',
 'host_type': 'domain',
 'path': '/123/abc',
 'query': '123',
 'fragment': 'abc'}
```
`IPvAnyAddress` would be similar.

**Comment by lilysikes** [comment]
So there isn't a way to do this with the json_encoder?

**Comment by StephenBrown2** [comment]
It's looking like it may be hard to do, as `HttpUrl` and `IPvAnyAddress` are subclasses of `str`, though based on the logic I see in the json_encoders, it should be catching it. I suspect it has to do with python's JSONEncoder [always checking if the element is an instance of `str`](https://github.com/python/cpython/blob/3.8/Lib/json/encoder.py#L191) and encoding based on that, but even subclassing HttpUrl and overriding the `__str__` method doesn't do it.

@samuelcolvin, shouldn't a custom definition for a class encoder override the default json encoding, even if it's a subtype of a native type? i.e. https://github.com/samuelcolvin/pydantic/blob/master/pydantic/main.py#L266?

**Comment by samuelcolvin** [comment]
Currently this isn't possible with `json_encoder`. You could achieve this by either:
* modifying the model's `dict` method to override the value of `url`
* or, if you don't want to modify `dict`, you could hack around withe `Config.json_dumps`

This should be possible in in future, either when I implement a fix for #935 or #951
